x-pack/plugins/actions/server/builtin_action_types/case/schema.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { schema } from '@kbn/config-schema';
+
+export const MappingActionType = schema.oneOf([
+  schema.literal('nothing'),
+  schema.literal('overwrite'),
+  schema.literal('append'),
+]);
+
+export const MapRecordSchema = schema.object({
+  source: schema.string(),
+  target: schema.string(),
+  actionType: MappingActionType,
+});
+
+export const CaseConfigurationSchema = schema.object({
+  mapping: schema.arrayOf(MapRecordSchema),
+});
+
+export const ExternalIncidentServiceConfiguration = {
+  apiUrl: schema.string(),
+  casesConfiguration: CaseConfigurationSchema,
+};
+
+export const ExternalIncidentServiceConfigurationSchema = schema.object(
+  ExternalIncidentServiceConfiguration
+);
+
+export const ExternalIncidentServiceSecretConfiguration = {
+  password: schema.string(),
+  username: schema.string(),
+};
+
+export const ExternalIncidentServiceSecretConfigurationSchema = schema.object(
+  ExternalIncidentServiceSecretConfiguration
+);
+
+export const UserSchema = schema.object({
+  fullName: schema.oneOf([schema.nullable(schema.string()), schema.maybe(schema.string())]),
+  username: schema.oneOf([schema.nullable(schema.string()), schema.maybe(schema.string())]),
```

Comment:
I made a note about the use of `schema.maybe()` in another comment, but this one seems ... more interesting.  I think technically you can just use `fullName: schema.nullable(schema.string())`, but am curious if you had to do it this way for some other reason.  Even for cases like `MappingActionType` above, I've instead sometimes just set the schema to a string, and validated the fixed set of literals in a custom validator, to produce a better error message for validation.

Probably worthwhile noting that the validation error messages produced from `schema.oneOf()` are quite verbose, and can be a little confusing.

Reply:
Not a good reason for that either. Probably I did it to match `fullname: string | undefined | null`. You are right, no need for that. 

About `schema.oneOf` you are right. It's quite verbose and difficult to catch. I tried to use the validate function but there was a bug (https://github.com/elastic/kibana/issues/64906). I would follow your advice and create a custom validator to produce better error messages for our schema in another PR.